### PR TITLE
chore(build): Add esm, cjs, umd & dist build targets

### LINKED ISSUE #11: Add build targets
Now that we've got #8 & #10 we should be good to start creating initial build targets for the package. And afterwards we can release an alpha version.

Targets we want are:
- ESM (ECMAScript modules) at `/lib/esm` for use in modern bundlers with tree-shaking
- CommonJS at `/lib/cjs` for use in Node
- UMD at `/lib/umd` for legacy bundlers that read one of AMD, CJS or browser global
- Browser global at `/dist` for folks that just want to include a `<script>` tag

After this (or maybe with it) we can do #2.

Check out [`react-composite-events`](https://github.com/benmvp/react-composite-events/blob/master/gulpfile.js) for an example of generating the build targets. I spent a lot of time researching and figuring it all out 😄

### PR DESCRIPTION
**Pull request by benmvp**
Added Gulp in order to create build targets for ECMAScript modules (ESM), CommonJS (cjs), Universal Modules (UMD), and dist bundles. This should hopefully be the final step before actually doing an inaugural release.

* ESM - used by modern dependency systems like Webpack or Rollup that can do tree-shaking optimizations
* CJS - used primarily by Node for resolving dependencies
* Dist bundle - used by legacy dependency systems like requireJS
* UMD - Can't think of a good use case where you'd use UMD over the previous 3, but adding it for now for completeness. May get removed in later releases

Was having issues with `lodash` including everything just for `isPlainObject` so I rewrote the code to no longer need it. In the future if we do need `lodash` we should use [`lodash-es`](https://www.npmjs.com/package/lodash-es) (lodash exported as ES modules).

Also changing the package name to be just `eventbrite` instead of `brite-rest`.

Fixes #11

### CODE REVIEW

File: src/request.ts [hunk 1]
```diff
@@ -53,8 +52,11 @@ export const fetchJSON = (
 };
 
 const hasArgumentsError = (responseData: JSONResponseData): boolean =>
-    isPlainObject(responseData['error_detail']) &&
-    isPlainObject(responseData['error_detail']['ARGUMENTS_ERROR']);
+    !!(
+        responseData &&
```
Comment (kaylieEB) [inline src/request.ts]: you don't need to check this with typescript right? i assume you'd get a warning otherwise since responseData is required

File: package.json [hunk 2]
```diff
@@ -30,12 +30,20 @@
...
-        "build:transpile": "babel src --ignore **/*.spec.ts --out-dir lib/cjs --extensions \".ts,.tsx\"",
+        "build:dist": "gulp build:dist build:dist:min",
+        "build:lib:cjs": "gulp build:lib:cjs",
+        "build:lib:esm": "gulp build:lib:esm",
+        "build:lib:umd": "gulp build:lib:umd build:lib:umd:min",
+        "prebuild:declarations": "rm -f lib/*.d.ts",
+        "prebuild:dist": "rm -rf dist",
+        "prebuild:lib:cjs": "rm -rf lib/cjs",
+        "prebuild:lib:esm": "rm -rf lib/esm",
+        "prebuild:lib:umd": "rm -rf lib/umd",
```
Comment (kaylieEB) [inline package.json]: I'm tempted to say that we should combine all the lib into one build:lib -- what would be the use case of wanting to build in one lib format but not the other? The time to build would be relatively incremental

File: gulpfile.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+const _getBabelStream = format =>
+  gulp
+    // get a stream of the files to transpile
+    .src(FILES_TO_BUILD)
+    // initialize the sourcemaps (used by UMD only)
+    .pipe(format === FORMAT_UMD ? sourcemaps.init() : util.noop())
+    // do the appropriate babel transpile (this is a copy from package.json)
+    .pipe(babel(_getBabelConfig(format)));
+
+const _genUmd = (minify = false) =>
```
Comment (kaylieEB) [inline gulpfile.js]: could we do ({minify = false}) instead? Boolean flags make it hard to follow below

File: gulpfile.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  _getBabelStream(FORMAT_UMD)
+    // If you're using UMD, you probably don't have `process.env.NODE_ENV` so, we'll replace it.
+    // If you're using the unminified UMD, you're probably in DEV
+    // If you're using the unminified UMD, you're probably in production
+    .pipe(
+      replace(
+        "process.env.NODE_ENV",
+        JSON.stringify(minify ? "production" : "development")
```
Comment (kaylieEB) [inline gulpfile.js]: actually instead you could set the parameter to `environment = 'development'` and then do `process.env.NODE_ENV || environment` right?

File: gulpfile.js (hunk 3 above)
Comment (benmvp) [reply gulpfile.js]: sure thing

File: gulpfile.js (hunk 4 above)
Comment (benmvp) [reply gulpfile.js]: Whoops! I totally forgot to add the conditional minification!

```
    .pipe(uglify())
    .pipe(rename({ extname: ".min.js" }))
```

I know that's not what you meant, but thanks for catching! :)

File: package.json (hunk 2 above)
Comment (benmvp) [reply package.json]: Yeah you're probably right. I'll think on it. I don't like the script explosion, but they're definitely gonna be separate in gulp. The kinda sucky part is that the declarations are also put in `lib/` so I can't just delete the whole `lib/` folder

File: src/request.ts (hunk 1 above)
Comment (benmvp) [reply src/request.ts]: So I don't need to check the existence of `responseData` because of TypeScript, but I do need to check the existence of of the sub properties since they are the response from the server. Will fix!

**Comment by benmvp** [comment]
@kaylieEB yes `lib/` & `dist/` were already in .gitignore: https://github.com/eventbrite/eventbrite-sdk-javascript/blob/master/.gitignore#L49-L51

**Comment by kwelch** [comment]
I have little gulp and no rollup experience. I will continue to peak at it throughout the day.

File: gulpfile.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            // don't place helpers at the top of the files, but point to reference contained external helpers
+            externalHelpers: true
+          },
+          _getBabelConfig(FORMAT_ESM)
+        )
+      ),
+
+      // Minify the code if that option is specified. `null` will get filtered out
+      // below
+      minify ? rollupUglify() : null
```
Comment (kaylieEB) [inline gulpfile.js]: can you do `minify && rollupUglify()`?

Comment (benmvp) [reply gulpfile.js]: Seems like that would work...

**Comment by benmvp** [comment]
@kwelch I'll probably merge the PR but it'll still be here! 😄

**Comment by kwelch-eb** [comment]
@benmvp I see no reason not to merge, I mostly want to learn more about it. I would love to hear why gulp? I would have expected rollup to handle all of these cases.

**Comment by benmvp** [comment]
@kwelch from my investigations Rollup is much like Webpack in that it's really good at taking all of the dependencies and creating a single bundle. Not so good at transpiling and generating target module formats for individual files, like we want for ESM, CJS & UMD.

That's where Gulp comes in. It's much better at taking a bunch of files and processing them all sorts of ways and outputting new versions of them in an efficient/parallel way.

So I guess when you break it down, Rollup is like `reduce` and Gulp is like `map`. 😂

File: package.json [hunk 6]
```diff
@@ -1,13 +1,13 @@
 {
-    "name": "brite-rest",
+    "name": "eventbrite",
```
Comment (rwholey-eb) [inline package.json]: why not something like eventbrite-sdk-javascript

File: gulpfile.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    globals: {
+      index: MODULE_NAME,
+      "isomorphic-fetch": "fetch"
+    },
+    exactGlobals: true
+  }
+];
+
+const _getBabelConfig = format => ({
+  babelrc: false,
```
Comment (rwholey-eb) [inline gulpfile.js]: is there too much dev stuff in there that we couldnt reuse it here?

File: gulpfile.js [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+      // Convert CommonJS modules to ES6 modules, so they can be included in a Rollup bundle
+      rollupCommonjs({
+        // Node modules are the ones we're trying to get it to understand
+        include: "node_modules/**"
+      }),
+
+      // Seamless integration between Rollup and Babel
+      rollupBabel(
+        Object.assign(
```
Comment (rwholey-eb) [inline gulpfile.js]: no spread?

File: package.json (hunk 6 above)
Comment (benmvp) [reply package.json]: Barrett wanted `eventbrite`. Having `javascript` in the name is pretty redundant considering it's a node module. And `sdk` is pretty much implied?

File: gulpfile.js (hunk 7 above)
Comment (benmvp) [reply gulpfile.js]: Yeah I think there's some sort of bug inheriting it here in gulp. I was trying to override the dev stuff and it just wasn't working, particularly the `modules: false` which we need for ESM

File: gulpfile.js (hunk 8 above)
Comment (benmvp) [reply gulpfile.js]: Cuz this is just running in Node, so didn't wanna assume Node > 8.X. Ideally folks can run the build on Node 6 as well. See: https://github.com/eventbrite/eventbrite-sdk-javascript/blob/master/.travis.yml#L2-L5

File: package.json (hunk 6 above)
Comment (kwelch-eb) [reply package.json]: I thought brite-rest was pretty funny. that being said my first attempt to import it I tried it as `eventbrite`

**Comment by kwelch-eb** [comment]
Thanks @benmvp! That is an amazing explanation.
